File: core/cyp/data/onchain.py
"""链上数据管线 stub（M3）：聪明钱/池深/持有分布/交易所净流。

无 RPC/Key 时返回 None（分析师降级），有注入 fetcher 时填充 OnchainData。
真实数据源（The Graph / Nansen / Dune 等）在实操阶段接入，本模块保证接口稳定。
"""

from __future__ import annotations

from cyp.contracts import OnchainData


class OnchainDataSource:
    """可注入 fetcher 的链上数据源：fetcher(symbol) -> dict | None。

    与 MarketSnapshot 组装方并列使用：
        onchain = await OnchainDataSource(fetcher).fetch(symbol)
        snap = MarketSnapshot(..., onchain=onchain)
    """

    def __init__(self, fetcher=None) -> None:
        self._fetcher = fetcher

    def is_configured(self) -> bool:
        return self._fetcher is not None

    async def fetch(self, symbol: str) -> OnchainData | None:
        if self._fetcher is None:
            return None   # 无数据源 → 分析师按 degraded 处理
        try:
            raw = await self._fetcher(symbol)
        except Exception:  # noqa: BLE001 —— 数据源失败隔离，降级不炸管线
            return None
        if not raw:
            return None
        return OnchainData(**raw)

File: internal/data/onchain.go
package data

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/wangchaozhi/cyp-agent/internal/contracts"
)

const maxOnchainResponseBytes = 1 << 20

// NewHTTPOnchainFetcher builds a read-only fetcher against an external
// onchain metrics API. The endpoint receives GET {base}?symbol=... and must
// answer with a JSON object matching contracts.OnchainData. Fetch failures
// degrade the onchain analyst instead of failing the snapshot.
func NewHTTPOnchainFetcher(baseURL string, client *http.Client) (OnchainFetcher, error) {
	parsed, err := url.Parse(strings.TrimSpace(baseURL))
	if err != nil || (parsed.Scheme != "http" && parsed.Scheme != "https") || parsed.Host == "" {
		return nil, fmt.Errorf("onchain data API must be an http(s) URL, got %q", baseURL)
	}
	if client == nil {
		client = &http.Client{Timeout: 10 * time.Second}
	}
	endpoint := *parsed
	return func(ctx context.Context, symbol string) (*contracts.OnchainData, error) {
		symbol = strings.TrimSpace(symbol)
		if symbol == "" {
			return nil, errors.New("onchain fetch requires a symbol")
		}
		target := endpoint
		query := target.Query()
		query.Set("symbol", symbol)
		target.RawQuery = query.Encode()

		request, err := http.NewRequestWithContext(ctx, http.MethodGet, target.String(), nil)
		if err != nil {
			return nil, err
		}
		request.Header.Set("Accept", "application/json")
		response, err := client.Do(request)
		if err != nil {
			return nil, err
		}
		defer func() { _ = response.Body.Close() }()
		if response.StatusCode != http.StatusOK {
			return nil, fmt.Errorf("onchain data API returned HTTP %d", response.StatusCode)
		}
		body, err := io.ReadAll(io.LimitReader(response.Body, maxOnchainResponseBytes))
		if err != nil {
			return nil, err
		}
		var payload contracts.OnchainData
		if err := json.Unmarshal(body, &payload); err != nil {
			return nil, fmt.Errorf("decode onchain payload: %w", err)
		}
		return &payload, nil
	}, nil
}

// OnchainEnrichedSource decorates a market data source with read-only onchain
// metrics. The underlying snapshot always wins: only a missing Onchain block
// is filled in, and fetch errors never fail the snapshot.
type OnchainEnrichedSource struct {
	inner   Source
	onchain *OnchainDataSource
}

func NewOnchainEnrichedSource(inner Source, onchain *OnchainDataSource) (*OnchainEnrichedSource, error) {
	if inner == nil {
		return nil, errors.New("onchain enrichment requires an inner market source")
	}
	return &OnchainEnrichedSource{inner: inner, onchain: onchain}, nil
}

func (source *OnchainEnrichedSource) Snapshot(
	ctx context.Context,
	symbol string,
) (contracts.MarketSnapshot, error) {
	snapshot, err := source.inner.Snapshot(ctx, symbol)
	if err != nil {
		return snapshot, err
	}
	if snapshot.Onchain == nil && source.onchain.IsConfigured() {
		snapshot.Onchain = source.onchain.Fetch(ctx, symbol)
	}
	return snapshot, nil
}
